src/compiler/scala/tools/nsc/util/ThreadIdentityAwareThreadLocal.scala

```diff
@@ -0,0 +1,39 @@
+package scala.tools.nsc.util
+
+object ThreadIdentityAwareThreadLocal {
+  def apply[T](valueOnWorker: => T) = new ThreadIdentityAwareThreadLocal[T](valueOnWorker, null.asInstanceOf[T])
+  def apply[T](valueOnWorker: => T,  valueOnMain: => T) = new ThreadIdentityAwareThreadLocal[T](valueOnWorker, valueOnMain)
+}
+
+// `ThreadIdentityAwareThreadLocal` allows us to have different (sub)type of values on main and worker threads.
```

Comment:
I think that we have 3 modes
admin - running setup and cleardown on the main thread
parallel-worker
single-worker

It may be easier to have this single worker on a seperate single thread to keep it simple

The ThreadFactory used for the async executor can provide a subclass of Thread - ParallelWorkerThread - it makes this switch easier